bentoml/_internal/io_descriptors/numpy.py

```diff
@@ -169,25 +179,29 @@ def _verify_ndarray(
         obj: "ext.NpNDArray",
         exception_cls: t.Type[Exception] = BadInput,
     ) -> "ext.NpNDArray":
+        import numpy as np
+
         if self._dtype is not None and self._dtype != obj.dtype:
-            if self._enforce_dtype:
-                raise exception_cls(
-                    f"{self.__class__.__name__}: enforced dtype mismatch"
-                )
-            try:
-                obj = obj.astype(self._dtype)  # type: ignore
-            except ValueError as e:
-                logger.warning(f"{self.__class__.__name__}: {e}")
+            # ‘same_kind’ means only safe casts or casts within a kind, like float64
+            # to float32, are allowed.
+            if np.can_cast(obj.dtype, self._dtype, casting="same_kind"):
+                obj = obj.astype(self._dtype, casting="same_kind")  # type: ignore
+            else:
+                msg = f'{self.__class__.__name__}: Expecting ndarray of dtype "{self._dtype}", but "{obj.dtype}" was received.'
+                if self._enforce_dtype:
+                    raise exception_cls(msg)
+                else:
+                    logger.warning(msg)
```

Comment:
This warning and the warning below can happen per request which can be very noisy.

Reply:
@ssheng should we just remove that log message for now?

Reply:
we can just make it a debug log.